File: java/harpc/src/main/java/com/bfd/harpc/loadbalance/common/DynamicHostSet.java
/**
 * Copyright (C) 2015 Baifendian Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.bfd.harpc.loadbalance.common;

import java.util.Collection;
import java.util.Set;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.bfd.harpc.common.ServerNode;
import com.google.common.collect.ImmutableSet;
import com.google.common.collect.Sets;

/**
 * 动态服务主机集合
 * <p>
 * 
 * @author : dsfan
 * @date : 2015-6-10
 */
public class DynamicHostSet implements DynamicSet<ServerNode> {

    /** LOGGER */
    private final Logger LOGGER = LoggerFactory.getLogger(getClass());

    /** 所有的{@link ServerNode} */
    private final Set<ServerNode> all = Sets.newHashSet();

    /** 可访问的{@link ServerNode} */
    private final Set<ServerNode> lives = Sets.newHashSet();

    /** 不可访问的{@link ServerNode} */
    private final Set<ServerNode> deads = Sets.newHashSet();

    /** {@link HostChangeMonitor} */
    private HostChangeMonitor<ServerNode> monitor;

    /**
     * add a server instance.
     * 
     * @param serverNode
     *            {@link ServerNode}
     */
    public synchronized void addServerInstance(ServerNode serverNode) {
        if (!lives.contains(serverNode)) {
            LOGGER.warn("add {} to lives set", serverNode);
            lives.add(serverNode);
            LOGGER.warn("add {} to all set", serverNode);
            all.add(serverNode);
            onChange();
        }
    }

    /**
     * add a live instance for heartbeat.
     * <p>
     * 
     * @param serverNode
     *            {@link ServerNode}
     */
    public synchronized void addLiveInstance(ServerNode serverNode) {
        if (all.contains(serverNode) && deads.contains(serverNode)) {
            lives.add(serverNode);
            LOGGER.warn("add {} to lives set", serverNode);
            deads.remove(serverNode);
            LOGGER.warn("remove {} from deads set", serverNode);
            onChange();
        }
    }

    /**
     * add dead instance
     * 
     * @param serverNode
     *            {@link ServerNode}
     */
    public synchronized void addDeadInstance(ServerNode serverNode) {
        if (all.contains(serverNode) && lives.contains(serverNode)) {
            deads.add(serverNode);
            LOGGER.warn("add {} to deads set", serverNode);
            lives.remove(serverNode);
            LOGGER.warn("remove {} from lives set", serverNode);
            if (lives.size() == 0) {
                adjustAll();
            }
            onChange();
        }
    }

    /**
     * 重新调整
     * <p>
     */
    private void adjustAll() {
        for (ServerNode serverNode : all) {
            lives.add(serverNode);
        }
        deads.clear();
    }

    /**
     * replace all hosts with new
     * 
     * @param hosts
     */
    public synchronized void replaceWithList(Collection<ServerNode> hosts) {
        String hostSetString = toString();
        // all替换为最新的集合
        all.clear();
        all.addAll(hosts);

        // deads取与最新集合的交集
        deads.retainAll(hosts);

        // lives取all和deads的差集
        lives.clear();
        lives.addAll(hosts);
        lives.removeAll(deads);

        LOGGER.warn("replace " + hostSetString + " with " + toString());
        onChange();
    }

    @Override
    public void monitor(HostChangeMonitor<ServerNode> monitor) {
        this.monitor = monitor;
        this.onChange();
    }

    /**
     * notify monitor
     * <p>
     */
    private void onChange() {
        if (monitor != null) {
            monitor.onChange(ImmutableSet.copyOf(lives));
        }
    }

    @Override
    public String toString() {
        return "[all=" + all + ", lives=" + lives + ", deads=" + deads + "]";
    }

    /**
     * getter method
     * 
     * @see DynamicHostSet#all
     * @return the all
     */
    public Set<ServerNode> getAll() {
        return all;
    }

    /**
     * getter method
     * 
     * @see DynamicHostSet#lives
     * @return the lives
     */
    public Set<ServerNode> getLives() {
        return lives;
    }

    /**
     * getter method
     * 
     * @see DynamicHostSet#deads
     * @return the deads
     */
    public Set<ServerNode> getDeads() {
        return deads;
    }
}

File: python/bfd/harpc/dynamic_host_set.py
# -*- coding: utf-8 -*-

# Copyright (C) 2015 Baifendian Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import socket
import logging

from bfd.harpc import settings
ASYNC_TAG = False


class DynamicHostSet(object):
    """dynamic host set"""
    section_name = "loadbalancer"

    def __init__(self, config):
        self._logger = logging.getLogger(__name__)
        self.all_nodes = set()
        self.live_nodes = set()
        self.dead_nodes = set()
        self._onchange = None
        if ASYNC_TAG:
            from gevent.lock import RLock
            self._lock = RLock()
        else:
            from threading import RLock
            self._lock = RLock()

        self._heartbeat_retry = config.getint(DynamicHostSet.section_name, "heartbeat_retry",
                                              default=settings.DEFAULT_HEARTBEAT_RETRY)
        self._heartbeat_time_out = config.getint(DynamicHostSet.section_name, "heartbeat_timeout", default=settings.DEFAULT_HEARTBEAT_TIMEOUT)

    def set_onchange(self, onchange):
        """set onchange func"""
        self._onchange = onchange

    def reset_with_list(self, node_list):
        """reset host set with given list"""
        with self._lock:
            self.all_nodes.clear()
            self.live_nodes.clear()
            self.dead_nodes.clear()
            self.all_nodes.update(node_list)
            self.live_nodes.update(node_list)
            self._onchange(list(self.live_nodes))

    def mark_dead(self, node):
        """mark a node dead"""
        with self._lock:
            if node in self.live_nodes:
                self.live_nodes.discard(node)
                self.dead_nodes.add(node)
                self._onchange(list(self.live_nodes))

    def recover_dead(self, node):
        """recover one dead node"""
        with self._lock:
            if node in self.all_nodes:
                if node in self.dead_nodes:
                    self._logger.info("recover dead nodes:%s" % node)
                    self.dead_nodes.discard(node)
                    self.live_nodes.add(node)
                    self._onchange(list(self.live_nodes))

    def recover_all(self):
        """recover all node"""
        with self._lock:
            self.live_nodes.update(self.dead_nodes)
            self.dead_nodes.clear()
            self._onchange(list(self.live_nodes))

    def heartbeat_check_all(self):
        """check all node state"""
        for node in self.all_nodes.copy():
            self.heartbeat_check(node)

    def heartbeat_check(self, node):
        """check node state"""
        addr = (node.split(":")[0], int(node.split(":")[1]))
        for i in range(settings.DEFAULT_HEARTBEAT_RETRY):
            s = socket.socket()
            s.settimeout(settings.DEFAULT_HEARTBEAT_TIMEOUT)
            try:
                s.connect(addr)
                self._logger.debug("heartbeat check node:%s success" % node)
                if node in self.dead_nodes:
                    self._logger.info("heartbeat check mark dead recover node:%s" % node)
                    self.recover_dead(node)
                break
            except socket.error as e:
                self._logger.warn("heartbeat check node:%s failed msg:%s " % (node, e))
                if i == settings.DEFAULT_HEARTBEAT_RETRY - 1:
                    if node in self.live_nodes:
                        self._logger.warn("heartbeat check mark dead add node:%s" % node)
                        self.mark_dead(node)
            finally:
                s.close()
